accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+
+#### What security benefits does Sigstore provide?
+[Sigstore](https://www.sigstore.dev/) is a collection of tools and services aimed at making it easy to use short-lived, ephemeral certificates to sign software.
+
+Sigstore has three main components: a CLI tool (Cosign), a Certificate Authority ([fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/)), and a time stamping and immutable ledger service ([rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/)).
+
+The Cosign functionality will be embedded directly in npm CLI, so we're left with the public [fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/) (CA) and [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/) (ledger) services.
+
+##### What does [fulcio.sigstore.dev](https://docs.sigstore.dev/fulcio/overview/) give us?
+- Independent party to validate claims of the OIDC identity token which contains references back to the repo, workflow run and git SHA.
+- A supported CI/CD provider can't falsify these claims and no one else can claim to be one of these CI/CD providers.
+- The npm build script can't forge the source provenance by, for example, saying it's from some other repo or git SHA.
+- Vendor-neutral: can support any standards-compliant OIDC Identity Provider that can meet their requirements.
+
+##### What does [rekor.sigstore.dev](https://docs.sigstore.dev/rekor/overview/) give us?
+- Verifies that the signing happened during the time the certificate was valid and that the response from [Rekor](https://docs.sigstore.dev/rekor/overview/) is signed and can be verified offline.
+- Verifiable, append-only ledger, makes it hard to tamper with already signed packages.
+- Verification won't rely on [Rekor](https://docs.sigstore.dev/rekor/overview/) as we'll store detached signatures on the registry for performance and reliability but it adds an extra layer of protection in case of registry compromise.
+
+#### How should maintainers authenticate to sign packages?
+[Sigstore](https://www.sigstore.dev/) issues short-lived “disposable keys” notarized against log-ins with OIDC identities. So what OIDC identity provider (IdP) do maintainers use?
+
+This has been the subject of some controversy; for instance, in the [proposal to add Sigstore signing to RubyGems](https://github.com/rubygems/rfcs/pull/37), users worried about “vendorization” (making the repository reliant on third parties to function, e.g. GitHub, Google or Microsoft) and privacy (some maintainers are pseudonymous; Sigstore certificates include email addresses by default when signing using the [cosign](https://github.com/sigstore/cosign) CLI tool).
+
+Sigstore supports both email and CI/CD workload-based OIDC identities:
+
+##### Email based identity provider
+For instance, Google, GitHub, Microsoft (supported by Sigstore today) or any other IdP that Sigstore might support.
```

Comment:
I know this is a problem introduced by proxy of using sigstore, but I'm uncomfortable with the likely future where maintainers who want to publish a package are relatively forced by their users to identify with a limited set of platforms that have extremely valid and well-documented criticisms around privacy (Google/Microsoft tracking, GitHub backing up on cookies policy),  politics (GitHub ICE, Microsoft DoD/JEDI deals, Google's Dragonfly, to name a few), and other ethical issues (treatment of women/Black/trans/other marginalized employees).

I don't fully believe that opting out of this feature for those reasons is a particularly equitable solution, nor one that will be an actual choice in the future.

Reply:
This is mentioned as the Vendorization problem. [RubyGems was the first to encounter this discussion](https://github.com/rubygems/rfcs/pull/37#issuecomment-1024591418) and a number of alternatives were floated both within the RubyGems discussion and in discussion with peers in other ecosystems (including npm).

1. Grow the list of endorsed IdPs. The current list isn't meant to be final, it's just what was convenient to bootstrap. There are many candidate IdPs who are non-commercial.
2. Operate an IdP specific to npm. This solves the vendorization problem (you've already decided to trust npm) but it almost completely cancels a critical advantage of independent IdPs, which is that to sign and upload an attacker will need to compromise two accounts.
3. A "neutral" IdP, independent of npm, probably shared among multiple ecosystem software repositories, operated by some trusted third party (at a first guess, the OpenSSF).
4. An email-based flow for verification. This has mostly been rejected as "OIDC over SMTP" -- a real possibility of protocol design flaws make this highly risky.

Reply:
I'm fine with #1 and maybe #3 (I fully understand why #2 and #4 are less than ideal). I'd really just want to see that list expanded somewhat significantly before this launches.

Reply:
We're sidestepping some of the problems with email based identities in this proposal by just supporting workflow identities from supported CI/CD systems. That said, we might wan to support email based identities in future if support improves.

Reply:
I certainly hope vendorization will be addressed from the beginning by offering alternative solutions. As other package managers run into the same issue, I see an opportunity for collaboration. If vendorization isn't addressed, you can expect a vocal group to resent this proposal, causing confusion and doubt and ultimately leading to failure of the whole thing. Improving supply chain security is too important to fail, so it should be done right to get everybody on board.

Reply:
For reference, [some further discussion of](https://github.com/sigstore/fulcio/issues/444) the Neutral IdP idea took place in Fulcio issue 444.

Reply:
> I certainly hope vendorization will be addressed from the beginning by offering alternative solutions.

💯 One of our internal goals of publishing this RFC is to help start these conversions with CI providers. I'm also excited about Sigstore as a solution to software signing for this reason. It's a vendor neutral entity that can help create alignment.

I'm not closed or opposed to extending npm package signing to support some kind of "author/maintainer" IdP in future that could allow maintainers to publish from their laptops. For the first iteration I think we'll start with only supporting build or workflow identities from CI/CD providers.